### PR TITLE
 (proof-new) SMT Preprocess proof generator

### PR DESCRIPTION
**Pull request by ajreynol**
This adds the proof generator for storing proofs of preprocessing.  It stores assertions added via preprocessing passes and their rewrites.  This utility will eventually live in SmtEngine.

It also adds 2 new proof rules, and fixes an issue in ProofNodeUpdater.

### CODE REVIEW

File: src/smt/preprocess_proof_generator.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+#include "expr/proof_node_manager.h"
+#include "theory/trust_node.h"
+
+namespace CVC4 {
+namespace smt {
+
+/**
+ * A proof generator storing proofs of preprocessing. This has two main
+ * interfaces during solving:
+ * (1) notifyNewAssert, for assertions that not part of the input and are
```
Comment (HanielB) [inline src/smt/preprocess_proof_generator.h:32]:
```suggestion
 * (1) notifyNewAssert, for assertions that are not part of the input and are
```

File: src/smt/preprocess_proof_generator.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  std::shared_ptr<ProofNode> getProofFor(Node f) override;
+  /** Identify */
+  std::string identify() const override;
+
+ private:
+  /** The proof node manager */
+  ProofNodeManager* d_pnm;
+  /**
+   * The trust node that was the source of each node constructed during
+   * preprocessing. For each n, d_src[n] is a trust node whose node is n. This
```
Comment (HanielB) [inline src/smt/preprocess_proof_generator.h:66]: ?
```suggestion
   * preprocessing. For each n, d_src[n] is a trust node whose node is n_src. This
```

File: src/smt/preprocess_proof_generator.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+   * Notify that n is a new assertion, where pg can provide a proof of n.
+   */
+  void notifyNewAssert(Node n, ProofGenerator* pg);
+  /**
+   * Notify that n was replaced by np due to preprocessing, where pg can
+   * provide a proof of the equality n=np.
+   */
+  void notifyPreprocessed(Node n, Node np, ProofGenerator* pg);
+  /** Get proof for */
+  std::shared_ptr<ProofNode> getProofFor(Node f) override;
```
Comment (HanielB) [inline src/smt/preprocess_proof_generator.h:57]: give the details of the structure of the proof node that is built for `f` in terms of transitivity.

File: src/smt/preprocess_proof_generator.h (hunk 2 above)
Comment (ajreynol) [reply src/smt/preprocess_proof_generator.h:66]: d_src is a data member of this class.

Comment (ajreynol) [reply src/smt/preprocess_proof_generator.h:66]: No, this is should say n.

File: src/smt/preprocess_proof_generator.h [hunk 4]
```diff
@@ -53,7 +53,11 @@ class PreprocessProofGenerator : public ProofGenerator
    * provide a proof of the equality n=np.
    */
   void notifyPreprocessed(Node n, Node np, ProofGenerator* pg);
-  /** Get proof for */
+  /**
+   * Get proof for, which returns a proof based on a proving an equality based
```
Comment (HanielB) [inline src/smt/preprocess_proof_generator.h:57]:
```suggestion
   * Get proof for f, which returns a proof based on proving an equality based
```

File: src/smt/preprocess_proof_generator.h (hunk 2 above)
Comment (HanielB) [reply src/smt/preprocess_proof_generator.h:70]: The confusing thing is that in the example below `n_src` is free

Comment (ajreynol) [reply src/smt/preprocess_proof_generator.h:70]: Haha, you want formal comments.  I can add "for some n_src".

Comment (HanielB) [reply src/smt/preprocess_proof_generator.h:70]: Ideally, yes! haha Other than that at least closed comments. :)

**Pull request merged**